executor/src/witgen/expression_evaluator.rs

```diff
@@ -55,47 +55,62 @@ where
-                if zero.constant_value().map(|z| z.is_zero()) == Some(true) =>
-            {
-                Ok(zero)
-            }
-            (Ok(left), op, Ok(right)) => match op {
-                AlgebraicBinaryOperator::Add => Ok(left + right),
-                AlgebraicBinaryOperator::Sub => Ok(left - right),
-                AlgebraicBinaryOperator::Mul => {
-                    if let Some(f) = left.constant_value() {
-                        Ok(right * f)
-                    } else if let Some(f) = right.constant_value() {
-                        Ok(left * f)
-                    } else {
-                        Err(IncompleteCause::QuadraticTerm)
+            AlgebraicBinaryOperator::Sub => Ok(self.evaluate(left)? - self.evaluate(right)?),
+            AlgebraicBinaryOperator::Mul => {
+                // don't short circuit on err as rhs might still be 0
+                let left_res = self.evaluate(left);
+                match left_res {
+                    Ok(left_expr) if left_expr.is_zero() => Ok(left_expr),
+                    Ok(left_expr) if left_expr.is_one() => self.evaluate(right),
+                    Ok(left_expr) => {
+                        if let Some(n) = left_expr.constant_value() {
+                            return Ok(self.evaluate(right)? * n);
+                        }
+                        // lhs not a constant
+                        let right_expr = self.evaluate(right)?;
+                        match right_expr.constant_value() {
+                            Some(n) if n.is_zero() => Ok(right_expr),
+                            Some(n) if n.is_one() => Ok(left_expr),
+                            Some(n) => Ok(left_expr * n),
+                            None => Err(IncompleteCause::QuadraticTerm),
+                        }
                     }
+                    // Err on lhs is ok if rhs is zero
+                    Err(left_err) => match self.evaluate(right) {
+                        Ok(right_expr) => {
+                            if let Some(n) = right_expr.constant_value() {
+                                if n.is_zero() {
+                                    return Ok(right_expr);
+                                }
+                            }
+                            Err(left_err)
+                        }
+                        Err(right_err) => Err(left_err.combine(right_err)),
+                    },
                 }
-                AlgebraicBinaryOperator::Pow => {
-                    if let (Some(l), Some(r)) = (left.constant_value(), right.constant_value()) {
-                        Ok(l.pow(r.to_integer()).into())
-                    } else {
-                        Err(IncompleteCause::ExponentiationTerm)
-                    }
+            }
+            AlgebraicBinaryOperator::Pow => {
+                if let (Some(l), Some(r)) = (
+                    self.evaluate(left)?.constant_value(),
+                    self.evaluate(right)?.constant_value(),
```

Comment:
It might be that `right` in this case could even always be a literal number - please check!

Reply:
added the `expect`